File: days/src/main/java/aoc2024j/day11/Day11.java

package aoc2024j.day11;

import utils.IO;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Day11 {

    record Stones(Map<Long, Long> stones) {
        Stones(String data) {
            this(Arrays
                    .stream(data.split(" "))
                    .collect(Collectors.groupingBy(Long::parseLong, Collectors.counting()))
            );
        }

        long numStones() {
            return stones.values().stream().mapToLong(Long::longValue).sum();
        }

        Stones blink(int steps) {
            var result = this;
            for (int i = 0; i < steps; i++) {
                result = result.blink();
            }
            return result;
        }

        private Stones blink() {
            var newStones =
                    stones.entrySet().stream()
                            .flatMap(entry -> {
                                var value = entry.getKey();
                                var count = entry.getValue();
                                return blinkValue(value).map(v -> Map.entry(v, count));
                            })
                            .collect(Collectors.groupingBy(Map.Entry::getKey,
                                    Collectors.summingLong(Map.Entry::getValue)));

            return new Stones(newStones);
        }

        private Stream<Long> blinkValue(long value) {
            if (value == 0) {
                return Stream.of(1L);
            } else {
                int digits = numberOfDigits(value);
                if (digits % 2 == 0) {
                    long div = pow10(digits / 2);
                    long high = value / div;
                    long low = value % div;
                    return Stream.of(high, low);
                } else {
                    return Stream.of(value * 2024);
                }
            }
        }

        private int numberOfDigits(long value) {
            int result = 1;
            while (value >= 10) {
                value /= 10;
                result++;
            }
            return result;
        }

        private long pow10(int exp) {
            long result = 1;
            for (int i = 0; i < exp; i++) {
                result *= 10;
            }
            return result;
        }
    }

    long part1(List<String> data) {
        var stones = new Stones(data.getFirst());
        var result = stones.blink(25);
        return result.numStones();
    }

    long part2(List<String> data) {
        var stones = new Stones(data.getFirst());
        var result = stones.blink(75);
        return result.numStones();
    }

    public static void main() {
        var day11 = new Day11();
        var data = IO.getResourceAsList("aoc2024/day11.txt");
        var part1 = day11.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day11.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day11/Day11.scala

package aoc2024s.day11;

import utils.IO

import scala.annotation.tailrec
import scala.jdk.CollectionConverters.*

object Day11 {

  case class Stones(stones: Map[Long, Long]) {

    def this(data: String) =
      this(
        data
          .split(" ")
          .groupMapReduce(_.toLong)(_ => 1L)(_ + _)
      )

    def numStones: Long = stones.values.sum

    def blink(steps: Int): Stones = {
      (1 to steps).foldLeft(this) { case (acc, _) =>
        acc.oneBlink
      }
    }

    private def oneBlink: Stones = {
      // I put the type annotation in the flatMap to select
      // the correct overload and get an iterable of tuples
      val newCounters =
        stones.flatMap[(Long, Long)] { (value, counter) =>
            blinkValue(value).map {
              _ -> counter
            }
          }
          .groupMapReduce(_._1)(_._2)(_ + _)

      Stones(newCounters)
    }

    private def numberOfDigits(value: Long): Int = {
      @tailrec
      def go(value: Long, acc: Int): Int =
        if value < 10 then acc
        else go(value / 10, acc + 1)

      go(value, 1)
    }

    private def pow10(exp: Int): Long = {
      (1 to exp)
        .foldLeft(1L) { case (acc, _) =>
          acc * 10
        }
    }

    private def blinkValue(value: Long): Seq[Long] =
      if value == 0 then Seq(1)
      else
        val n = numberOfDigits(value)
        if n % 2 == 0 then
          val div = pow10(n / 2)
          val left = value / div
          val right = value % div
          Seq(left, right)
        else
          Seq(value * 2024)
  }

  def part1(data: List[String]): Long = {
    val stones = new Stones(data.head)
    val result = stones.blink(25)
    result.numStones
  }

  def part2(data: List[String]): Long = {
    val stones = new Stones(data.head)
    val result = stones.blink(75)
    result.numStones
  }

  @main def main11(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day11.txt").asScala.toList;
    val part1 = Day11.part1(data);
    println(s"part1 = $part1");
    val part2 = Day11.part2(data);
    println(s"part2 = $part2");
  }
}